gfan: fails to build: setoper.h: No such file or directory

**Describe the bug**

Building `/nix/store/9mi04kgd0l886j0jbwr16c7fx2pmf5pf-gfan-0.6.2.drv` fails with the following messages:

```shell
$ nix-build -A gfan
these derivations will be built:
  /nix/store/9mi04kgd0l886j0jbwr16c7fx2pmf5pf-gfan-0.6.2.drv
building '/nix/store/9mi04kgd0l886j0jbwr16c7fx2pmf5pf-gfan-0.6.2.drv'...
unpacking sources
unpacking source archive /nix/store/a343sk3xjzjwf97d6sc70d53qi5hzryk-gfan0.6.2.tar.gz
source root is gfan0.6.2
setting SOURCE_DATE_EPOCH to timestamp 1506606252 of file gfan0.6.2/Makefile
patching sources
configuring
no configure script, doing nothing
building
build flags: SHELL=/nix/store/k8p54jg8ipvnfz435mayf5bnqhw4qqap-bash-4.4-p23/bin/bash CC=cc CXX=c++ cddnoprefix=1
c++ -DGMPRATIONAL -Wuninitialized -fno-omit-frame-pointer -O2     -g       -std=c++0x -g -DNOCDDPREFIX  -c src/polynomialgcd.cpp -o src/polynomialgcd.o
c++ -DGMPRATIONAL -Wuninitialized -fno-omit-frame-pointer -O2     -g       -std=c++0x -g -DNOCDDPREFIX  -c src/lp_cdd.cpp -o src/lp_cdd.o
src/lp_cdd.cpp:4:10: fatal error: setoper.h: No such file or directory
    4 | #include "setoper.h"
      |          ^~~~~~~~~~~
compilation terminated.
make: *** [Makefile:549: src/lp_cdd.o] Error 1
builder for '/nix/store/9mi04kgd0l886j0jbwr16c7fx2pmf5pf-gfan-0.6.2.drv' failed with exit code 2
error: build of '/nix/store/9mi04kgd0l886j0jbwr16c7fx2pmf5pf-gfan-0.6.2.drv' failed
```

Note that `git bisect` points to https://github.com/NixOS/nixpkgs/commit/22212571164eff6af18c5dec90ecf17cdc1387ce which updates `cddlib`.

**To Reproduce**
Steps to reproduce the behavior:
1. `nix-build -A gfan`

**Expected behavior**
The package is built successfully.

**Notify maintainers**
@7c6f434c (gfan, cddlib), @timokau (cddlib)

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.70-hardened, NixOS, 20.03.3133.3a10a004bb5 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`
 - channels(root): `"nixos-20.03.3133.3a10a004bb5"`
 - channels(omasanori): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - nixpkgs.gfan
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/applications/science/math/gfan/default.nix
+++ b/pkgs/applications/science/math/gfan/default.nix
@@ -8,11 +8,15 @@ stdenv.mkDerivation rec {
     sha256 = "02pihqb1lb76a0xbfwjzs1cd6ay3ldfxsm8dvsbl6qs3vkjxax56";
   };
 
-  patchPhase = stdenv.lib.optionalString stdenv.cc.isClang ''
+  patches = [
+    ./gfan-0.6.2-cddlib-prefix.patch
+  ];
+
+  postPatch = stdenv.lib.optionalString stdenv.cc.isClang ''
     substituteInPlace Makefile --replace "-fno-guess-branch-probability" ""
   '';
 
-  buildFlags = [ "CC=cc" "CXX=c++" "cddnoprefix=1" ];
+  buildFlags = [ "CC=cc" "CXX=c++" ];
   installFlags = [ ''PREFIX=$(out)'' ];
   buildInputs = [ gmp mpir cddlib ];
 

--- a/pkgs/applications/science/math/gfan/gfan-0.6.2-cddlib-prefix.patch
+++ b/pkgs/applications/science/math/gfan/gfan-0.6.2-cddlib-prefix.patch
@@ -1,0 +1,55 @@
+diff -ru gfan0.6.2.orig/src/app_librarytest.cpp gfan0.6.2/src/app_librarytest.cpp
+--- gfan0.6.2.orig/src/app_librarytest.cpp	2020-10-19 08:41:27.981863500 +0900
++++ gfan0.6.2/src/app_librarytest.cpp	2020-10-19 08:42:44.551863500 +0900
+@@ -12,8 +12,8 @@
+ #include "setoper.h"
+ #include "cdd.h"
+ #else
+-#include "cdd/setoper.h"
+-#include "cdd/cdd.h"
++#include "cddlib/setoper.h"
++#include "cddlib/cdd.h"
+ #endif
+ #include <iostream>
+ #include <fstream>
+diff -ru gfan0.6.2.orig/src/gfanlib_zcone.cpp gfan0.6.2/src/gfanlib_zcone.cpp
+--- gfan0.6.2.orig/src/gfanlib_zcone.cpp	2020-10-19 08:41:27.981863500 +0900
++++ gfan0.6.2/src/gfanlib_zcone.cpp	2020-10-19 08:42:44.571863500 +0900
+@@ -16,8 +16,8 @@
+ #include "setoper.h"
+ #include "cdd.h"
+ #else
+-#include "cdd/setoper.h"
+-#include "cdd/cdd.h"
++#include "cddlib/setoper.h"
++#include "cddlib/cdd.h"
+ #endif
+ //}
+ 
+@@ -52,8 +52,8 @@
+ 				  "dd_free_global_constants()\n"
+ 				  "in your deinitialisation code (only available for cddlib version>=094d).\n"
+ 				  "This requires the header includes:\n"
+-				  "#include \"cdd/setoper.h\"\n"
+-				  "#include \"cdd/cdd.h\"\n"
++				  "#include \"cddlib/setoper.h\"\n"
++				  "#include \"cddlib/cdd.h\"\n"
+ 				  "\n"
+ 				  "Alternatively, you may call gfan:initializeCddlibIfRequired() and deinitializeCddlibIfRequired()\n"
+ 				  "if gfanlib is the only code using cddlib. If at some point cddlib is no longer required by gfanlib\n"
+diff -ru gfan0.6.2.orig/src/lp_cdd.cpp gfan0.6.2/src/lp_cdd.cpp
+--- gfan0.6.2.orig/src/lp_cdd.cpp	2020-10-19 08:41:27.991863500 +0900
++++ gfan0.6.2/src/lp_cdd.cpp	2020-10-19 08:42:44.571863500 +0900
+@@ -5,9 +5,9 @@
+ #include "cdd.h"
+ #include "cdd_f.h"
+ #else
+-#include "cdd/setoper.h"
+-#include "cdd/cdd.h"
+-#include "cdd/cdd_f.h"
++#include "cddlib/setoper.h"
++#include "cddlib/cdd.h"
++#include "cddlib/cdd_f.h"
+ #endif
+ //}
+ #include "termorder.h"

